### PR TITLE
Private visibility is toxic to extensibility

### PR DESCRIPTION
**Pull request by dhensby**
See #3874

### CODE REVIEW

**Comment by MatthewBonner** [comment]
I can't believe you have done this, surely there should be some sort of discussion around what kind of re factoring needs to take place? For instance, the properties you have made protected should be private, and you should either have protected or public accessors and mutators so that the class cannot be changed or the type if the property contains a scalar value.

**Comment by micmania1** [comment]
I think more consideration needs to be given to what is being changed and why. I don't think changing every single instance is beneficial.

For example, I'm not sure its even possible to extend Config_LRU or Injector.

Also, because they're private, its much safer to remove or change their function in future versions because there is no chance that they're being used elsewhere in bespoke code or modules (or even in the framework).

**Comment by dhensby** [comment]
> there should be some sort of discussion around what kind of re factoring needs to take place?

Yes, this PR is to instigate that

> the properties you have made protected should be private, and you should either have protected or public accessors and mutators so that the class cannot be changed or the type if the property contains a scalar value

Can you give a specific example? Why should they be private specifically and other properties not? What's so special about the properties I've changed?

@micmania1 I agree with you in general terms. In fact, [Symfony 2 recently changed all their protected properties to private](http://fabien.potencier.org/article/47/pragmatism-over-theory-protected-vs-private) but, as that article states:

> Having a few well defined extension points force the developer to extend your library the right way instead of hacking your code.


My main aim with this PR is to bring in some consistency - a better solution can be found for master or when we're somewhere closer to a level that Symfony is.

**Comment by dhensby** [comment]
> I'm not sure its even possible to extend Config_LRU or Injector

Injector definitely can: `Injector::set_inst(new MyExtendedInjector());`

**Comment by MatthewBonner** [comment]
There are a few bad practises with the following example, but it highlights that complex code can lead to undesirable results, which may be hard to debug, this is only a simple example, if I was to replace it with some code from SS then the debugging would become a real nightmare.

```php
<?php

class Task {
  private $ran = FALSE;

  public function hasRan () {
    return $this->ran;
  }

  public function init () {
    set_memory_limit(-1);
  }

  public function run () {
    // run task
  }
}

class TaskWithNoInit {
  private $ran = FALSE;

  public function hasRan () {
    return $this->ran;
  }

  public function run () {
    // run task
  }
}

class Parent {
  protected $task;

  public function __construct () {
    $this->task = new Task;
  }

  protected function ensureInitLogic () {
    if ($this->task->hasRan()) {
      $this->task->init();
    }
  }

  public function doLogic () {
    $this->task->run();
  }
}

class Child {
  protected function ensureInitLogic () {
    if ($this->task->hasRan()) {
      $this->task = new TaskWithNoInit;
      parent::ensureInitLogic();
    }
  }
}

?>
```

**Comment by MatthewBonner** [comment]
The following shows a way to solve the problem, making it easier to debug, and prevent programmers from making silly mistakes...

```php
<?php

interface ITask {
  function hasRan();
  function run();
}

class Task implements ITask {
  private $ran = FALSE;

  public function hasRan () {
    return $this->ran;
  }

  public function init () {
    set_memory_limit(-1);
  }

  public function run () {
    // run task
  }
}

class TaskWithNoInit implements ITask {
  private $ran = FALSE;

  public function hasRan () {
    return $this->ran;
  }

  public function run () {
    // run task
  }
}

class Parent {
  private $task;

  public function __construct () {
    // This poorly designed class requires the specific Task class, opposed to a class implementing the ITask interface
    $this->task = new Task;
  }

  protected function getTask () {
    return $this->task;
  }

  protected function ensureInitLogic () {
    if ($this->task->hasRan()) {
      $this->task->init();
    }
  }

  public function doLogic () {
    $this->task->run();
  }
}

class Child {
  protected function ensureInitLogic () {
    if ($this->getTask()->hasRan()) {
      // This won't work, because $this->task does not exist
      $this->task = new TaskWithNoInit;
      parent::ensureInitLogic();
    }
  }
}

?>

**Comment by MatthewBonner** [comment]
To answer the following question:
"Why should they be private specifically and other properties not? What's so special about the properties I've changed?"

Realistically, in my view, all properties should be private due to PHP's lack of features in making properties read-only, and due to the type juggling. If we enforce stricter rules then the code as a result will naturally become cleaner and more robust.

**Comment by dhensby** [comment]
@MatthewBonner your example code is very hard to follow because you aren't explicitly defining inheritance and there are no comments to highlight in what way this relates to any of the changes in my PR.

For arguments sake I present #3869 

If you want to extend SS_Query to query against a standard array then the access to the private properties is ESSENTIAL. A simple example below:

```php

class ArrayQuery extends SS_Query {

    protected $record = array();

    public function nextRecord() {
        return isset($this->record[$this->rowNumber + 1]) ? $this->record[$this->rowNumber + 1] : false;
    }

}

```

This could be done with protected or public accessor mthods but that's just adding another layer... Ok, if the class had those methods, then I'd not need to change the visibility - but I don't see an inherent advantage in making them private.

I do suppose it means the ability to more easily remove those properties late for some other ones, but I guess that's a massive refactor job for the future and not inline with current conventions

**Comment by dhensby** [comment]
> To answer the following question:
> "Why should they be private specifically and other properties not? What's so special about the properties I've changed?"
> 
> Realistically, in my view, all properties should be private due to PHP's lack of features in making properties read-only, and due to the type juggling. If we enforce stricter rules then the code as a result will naturally become cleaner and more robust.

Ok, so the real crux here is you think all properties should be `private`? It's not that I'm changing specific properties that you feel are somewhat special.

If that's the case, then you're welcome to take on that responsibility and start putting that against master.

But, in the absence of a convention that says all properties should be private, this is a solution that brings the few exceptions within the framework in-line with the rest of the code base until a better and more robust solution is implemented.

**Comment by MatthewBonner** [comment]
What is wrong with...

```php
<?php

class SS_Query {
  private $record;

  protected function getRecord () {
    return $this->record;
  }
}

?>
```

**Comment by dhensby** [comment]
Well, your solution misses the point and also `$record` shouldn't be inherently on an `SS_Query`... it's meant to be abstract to the storage method (array, DB query, API result set, etc)

With your example accessing `$this->rowNumber` still throws a fatal error for accessing a private property on a parent class...

I don't disagree that adding a layer of functions to get the property value is a better solution, but it's just a significantly more difficult and disruptive change than this one

**Comment by MatthewBonner** [comment]
The example was based on the code in this context, not in the context of SilverStripe, for the sake of simplicity.

You have to weigh up the breaking changes in comparison with making a change that requires more effort in the short-term.

Just because it is easier to make private properties protected, it doesn't mean it is the right way to solve the problem of extending the API.

**Comment by dhensby** [comment]
This PR is in the context of SS, so we should keep our discussion of the PR in that context too.

> Just because it is easier to make private properties protected, it doesn't mean it is the right way to solve the problem of extending the API.

Agreed, but this is a fix that fits with the current conventions and consistency is better than inconsistency.

**Comment by MatthewBonner** [comment]
Well at some point I've got to lead the upgrade of a SS 3.1 site to SS 3.2 when it is released, due to the size of the website it has taken over a year to upgrade from SS 2 to SS 3 and then implement a new theme. We've had countless problems with sloppy code, causing hours of debugging nightmares and causing progress to slow, when realistically the upgrade process should have been relatively smooth. Your proposing making a change that is eventually going to lead to having to make breaking changes, and it is this disregard to breaking changes in SS that I really despise, it makes my job, and the job of others incredibly tedious when it doesn't need to be. All we have to do is follow some simple rules and people will be amazed at the quality we are churning out.

I wouldn't argue about one or two protected properties per class, but making them all protected is a mistake I made some time ago, and rather than you having to go through the same learning curve on why this is a bad idea, I'd prefer to encourage you to avoid doing such a thing in the hope that I can save you and others the same pain.

**Comment by MatthewBonner** [comment]
I've updated my example to keep it in context with SilverStripe.

**Comment by camfindlay** [comment]
@dhensby probably something to raise at the next core committers hangout too after a good discussion in the dev list perhaps to thrash out the pros/cons of the change.

**Comment by sunnysideup** [comment]
+1 on the approach below (and above)  - BUT where it gets really powerful is shown in method doSomethingElse, because this is where you can influence every use of records within a framework of classes in a pretty efficient way (by overriding the parent getRecords methods in a child class): 
```php


<?php

class SS_Query {
  private $record;
  private $rowNumber ;

  protected function getRecord() {
    return $this->record;
  }

  protected function getRowNumber() {
    return $this->rowNumber;
  }
 
 private function doSomethingElse(){
  //NOTE - we are NOT accessing the variable directly here.
  $myRecords = $this->getRecord();
 }
}

class ArrayQuery extends SS_Query {
    public function nextRecord() {
        $record = $this->getRecord();
        $rowNumber = $this->getRowNumber();
        return isset($record[$rowNumber + 1]) ? $record[$rowNumber + 1] : false;
    }
}

?>

**Comment by tractorcow** [comment]
The point I'd like to make that, any class should be extendible, either vertically (by subclassing) or to a more restricted degree, horizontally (by adding extensions). A good framework should respect the open closed principle, meaning that it should be able to have its behaviour overridden without having to edit internal code.

Either of the above approaches are reasonable here, although we could definitely have another discussion about which getters are appropriate as protected or public.

I personally prefer having getters, and enforcing usage of these over directly accessing private variables, so I'm leaning towards the solution put forward by Matt, but we would have to be more thorough in our review of each usage of private variables in each class.

Whatever solution we take needs to support the mechanisms and coding practices we've developed for SilverStripe that might not apply to other frameworks (e.g. extensions).

So, let's take it to the devlist and see what the community prefer.

**Comment by tractorcow** [comment]
Did someone add a reply then delete it? :P

**Comment by sminnee** [comment]
Hey @dhensby do you think there is value in keeping this PR open? I'm not sure if it's mergeable as-is and I'm not sure what the path forward would be. It's stalled for a year...

**Comment by tractorcow** [comment]
I think we can close this, but only if @dhensby is happy for us to do so.

In future, maybe we just add accessors where needed, rather than converting member visibility.

**Comment by sminnee** [comment]
Yeah my main concern with "everything protected" is that it substantially increases our public API, which limits what we can do in minor versions. I guess we can say "protected methods are not part of the public API" but then we're basically saying that anyone who uses this extension-point is in no-mans land, which isn't great either.

**Comment by dhensby** [comment]
I think either we make them all private and have a well defined API or we open them up and don't.

As opening them up and having a badly defined API is only going to cause us problems, I think a better solution is to add setter/getter methods for these
